raft: don't cap commit index in leader heartbeats

lyang24 opened the pull request:
This is a follow up to #122690 which makes raft leader to send the most up-to-date commit index in heartbeats, instead of capping it at the follower's `Match` index. The follower can safely process commit indices that are out of its log bounds.

CockroachDB version gating infrastructure is added to raft to ensure no mixed heartbeat behaviour during cluster upgrade. If a leader running the new code sends a high commit index to a follower running the old code, the follower would crash. The version gate makes sure all followers are upgraded to the new behaviour first.

Resolves #122100

lyang24 committed:
raft: don't cap commit index in leader heartbeats

This is a follow up to d6da6c6 which makes raft leader to send the most
up-to-date commit index in heartbeats, instead of capping it at the
follower's Match index. The follower can safely process commit indices
that are out of its log bounds.

CockroachDB version gating infrastructure is added to raft to ensure no
mixed heartheat behaviour during cluster upgrade. If a leader running
the new code sends a high commit index to a follower running the old
code, the follower would crash. The version gate makes sure all
followers are upgraded to the new behaviour first.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124648)
<!-- Reviewable:end -->

pav-kv reviewed (approved):
LGTM % nits

Something's wrong with CI, try to rebase after changes.

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124648)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/clusterversion/cockroach_versions.go` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124648#gh-1627480800:-NzdMl4f7eiywp6h5BWN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6294cb02fa65657f6e4153cdfbf28ebf7cb40792/pkg/clusterversion/cockroach_versions.go#L256)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

```suggestion
	// V24_2_FlexibleRaftFollowerCommitIndex allows raft leader to send its commit
	// index in a heartbeat without capping it at the follower's Match index.
```
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124648#gh-1627505110:-NzdMmjA-BNuy1kNxeP7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6294cb02fa65657f6e4153cdfbf28ebf7cb40792/pkg/raft/raft.go#L266)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

```suggestion
	// CRDBVersion gives access to the active version of CockroachDB, which allows
	// implementing version gates for making backward incompatible changes safely.
```
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 693 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124648#gh-1627519111:-NzdMnciBLWqW0GxioNw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6294cb02fa65657f6e4153cdfbf28ebf7cb40792/pkg/raft/raft.go#L693)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`context.Background()` should be good
</blockquote></details>

Done.
___
*[`pkg/raft/log.go` line 0 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124648#gh-1627487949:-NzdMoiMA7FyUFu11mar:bvth56w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6294cb02fa65657f6e4153cdfbf28ebf7cb40792/pkg/raft/log.go#L0)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/cockroachdb/cockroach/pull/124648#discussion_r1627487949) to the right spot — sorry!)*

Let's leave this check as a defence-in-depth. It won't be triggered now though.
</blockquote></details>

reverted


<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -670,6 +678,11 @@ func (r *raft) sendHeartbeat(to uint64) {
 	// The leader MUST NOT forward the follower's commit to
 	// an unmatched index.
 	commit := min(pr.Match, r.raftLog.committed)
+	// If all the cluster nodes are on the new version
+	// then we can safely use leader's commit index in heartbeat.
+	if r.crdbVersion != nil && r.crdbVersion.IsActive(context.Background(), clusterversion.V24_2_DisableRaftCommitIndexCapping) {
+		commit = r.raftLog.committed
+	}
```
Can we flip this to:

```golang
commit := r.raftLog.committed

if version != nil && !version.IsActive(...) {
	commit = min(commit, pr.Match)
}
```
and rearrange the comments correspondingly?

The rationale is:
- the conditional will be simply removed in the future
- we want to test the new behaviour if the version handle is nil (in some tests)

Also, this allows removing `cluster.MakeTestingClusterSettingsWithVersions` from `raft_test.go`, as the new behaviour works in tests by default with a nil.

I think once we flip this code, we should see an (expected) change in the `lagging_commit.txt` test.

lyang24 reviewed (commented):
> Reviewed 1 of 8 files at r1, 1 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124648)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/raft.go` line 685 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124648#gh-1628149521:-NzdW1Hg3soJIQlkXhJM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98067d48a340f34018a92156f8d59699bcacae72/pkg/raft/raft.go#L685)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Can we flip this to:

```golang
commit := r.raftLog.committed

if version != nil && !version.IsActive(...) {
	commit = min(commit, pr.Match)
}
```
and rearrange the comments correspondingly?

The rationale is:
- the conditional will be simply removed in the future
- we want to test the new behaviour if the version handle is nil (in some tests)

Also, this allows removing `cluster.MakeTestingClusterSettingsWithVersions` from `raft_test.go`, as the new behaviour works in tests by default with a nil.

I think once we flip this code, we should see an (expected) change in the `lagging_commit.txt` test.
</blockquote></details>

Done.
___
*[`pkg/raft/raftpb/confstate.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124648#-NzdwFqgAuH-xnxktq7l:-NzdwFqgAuH-xnxktq7m:b-z66c0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/51614ccb781f033eb3b1d921cefe90924e42a925/pkg/raft/raftpb/confstate.go#L46)):*
> ```Go
> }
> 
> func DescribeConfState(state ConfState) string {
> ```

move DescribeConfState here to avoid cyclic dependency in pkg raft_pb 


<!-- Sent from Reviewable.io -->

pav-kv committed:
raft,roachpb: move DescribeConfState to raftpb

This removes the unnecessary direct dependency of roachpb package on
raft package.

Epic: none
Release note: none

pav-kv commented on pkg/raft/testdata/lagging_commit.txt:
```diff
@@ -134,7 +134,7 @@ process-ready 1
-1->3 MsgHeartbeat Term:1 Log:0/0 Commit:11
+1->3 MsgHeartbeat Term:1 Log:0/0 Commit:13
 
 # Since the heartbeat message does not bump the follower's commit index, it will
```
This comment now needs updating, since we fixed the issue that it describes: we can see that the entries are now committed earlier by one round of communication.

pav-kv replied:
Done.

nvanbenschoten commented on pkg/raft/raft.go:
```diff
@@ -663,13 +671,18 @@ func (r *raft) maybeSendSnapshot(to uint64, pr *tracker.Progress) bool {
 	// The leader MUST NOT forward the follower's commit to
 	// an unmatched index.
-	commit := min(pr.Match, r.raftLog.committed)
+	if r.crdbVersion != nil && !r.crdbVersion.IsActive(context.Background(), clusterversion.V24_2_DisableRaftCommitIndexCapping) {
```
Do we want to find a way to test this? Maybe we add a mixed-version variant of `lagging_commit.txt`?

lyang24 replied:
I am trying to see what is the best way to add the mixed version tests
I am thinking adding parsing logic to assign raft's crdb cluster version handle base on input at
`pkg/raft/rafttest/interaction_env_handler.go`

am i in the ball park? cc @pav-kv

pav-kv replied:
Yeah, you could extend the `add-nodes` command with a version parameter. Then, in the test, instead of a single `add-nodes 3` command (which adds 3 nodes configured the same way), you would need to `add-nodes 1` configured one way, and `add-nodes 2` configured the other way.

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124648)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @pav-kv)

___
*[`pkg/raft/testdata/lagging_commit_mixed_cluster_version.txt` line 50 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/124648#-O-YKSbF3pGycwQwSV9u:-O-YKSbF3pGycwQwSV9v:b-w2hmw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eca79e30eb4849f5d22a8e93cb4927d71c340da/pkg/raft/testdata/lagging_commit_mixed_cluster_version.txt#L50)):*
> ```help
> 1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "data1"]
> 1->2 MsgApp Term:1 Log:1/12 Commit:11 Entries:[1/13 EntryNormal "data2"]
> 1->3 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "data1"]
> ```

Hey @pav-kv , I looked into adding nodes separately with different cluster versions. I added two nodes with older version and one node with dont cap commit index active as node 3. But it seems like the leader node 1 cannot talk to 3 for some reason as this test will fail here only showing the logs of node 1 send msg to node 2. Do you happen to know what i did wrong? 


<!-- Sent from Reviewable.io -->
